triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala

```diff
@@ -392,40 +389,40 @@ class Runner(
-          val messageVal = message match {
-            case TransactionMsg(transaction) => {
-              converter.fromTransaction(transaction) match {
-                case Left(err) => throw new ConverterException(err)
-                case Right(x) => x
-              }
+      .toMat(Sink.fold[SValue, TriggerMsg](evaluatedInitialState)((state, message) => {
+        val messageVal = message match {
+          case TransactionMsg(transaction) => {
+            converter.fromTransaction(transaction) match {
+              case Left(err) => throw new ConverterException(err)
+              case Right(x) => x
             }
-            case CompletionMsg(completion) => {
-              val status = completion.getStatus
-              if (status.code != 0) {
-                logger.warn(s"Command failed: ${status.message}, code: ${status.code}")
-              }
-              converter.fromCompletion(completion) match {
-                case Left(err) => throw new ConverterException(err)
-                case Right(x) => x
-              }
+          }
+          case CompletionMsg(completion) => {
+            val status = completion.getStatus
+            if (status.code != 0) {
+              logger.warn(s"Command failed: ${status.message}, code: ${status.code}")
+            }
+            converter.fromCompletion(completion) match {
+              case Left(err) => throw new ConverterException(err)
+              case Right(x) => x
             }
-            case HeartbeatMsg() => converter.fromHeartbeat
           }
-          val clientTime: Timestamp =
-            Timestamp.assertFromInstant(Runner.getTimeProvider(timeProviderType).getCurrentTime)
-          machine.setExpressionToEvaluate(
-            makeApp(
-              update,
-              Array(SEValue(STimestamp(clientTime)): SExprAtomic, SEValue(messageVal), state)))
-          val value = Machine.stepToValue(machine)
-          val newState = handleStepResult(value, submit)
-          SEValue(newState)
-        }))(Keep.right[NotUsed, Future[SExpr]])
+          case HeartbeatMsg() => converter.fromHeartbeat
+        }
+        val clientTime: Timestamp =
+          Timestamp.assertFromInstant(Runner.getTimeProvider(timeProviderType).getCurrentTime)
+        machine.setExpressionToEvaluate(
+          makeApp(update, Array(STimestamp(clientTime): SValue, messageVal, state)))
+        val value = Machine.stepToValue(machine)
+        val newState = handleStepResult(value, submit)
+        newState
+      }))(Keep.right[NotUsed, Future[SValue]])
   }
 
-  def makeApp(func: AExpr, args: Array[SExprAtomic]): AExpr = {
+  def makeApp(func: AExpr, values: Array[SValue]): AExpr = {
+    val args: Array[SExprAtomic] = values.map(SEValue(_))
+    // We can safely introduce a let-expression here to bind the `func` expression,
+    // because there are no stack-references in `args`, since they are pure speedy values.
```

Comment:
Given that the coding is a lot cleaner now, I wouldn't have asked for the comment anymore. 😃
But it doesn't do harm either.